### ISSUE TITLE
Clingo overriding packages.yaml variant preferences

### ISSUE THREAD

**Issue by adamjstewart** [LLNL/spack#24754]
I'm trying to install PyTorch without CUDA support. With a fresh clone of Spack and no `packages.yaml`, I see:
```console
$ spack solve py-torch
==> Best of 9 considered solutions.
==> Optimization Criteria:
  Priority  Criterion                                 Value
  1         deprecated versions used                      0
  2         version weight                                0
  3         number of non-default variants (roots)        1
  4         multi-valued variants                        -1
  5         preferred providers for roots                 0
  6         number of non-default variants (non-roots)    1
  7         preferred providers (non-roots)             130
  8         compiler mismatches                           0
  9         version badness                              60
  10        count of non-root multi-valued variants      -5
  11        non-preferred compilers                       0
  12        target mismatches                             0
  13        non-preferred targets                         0

py-torch@1.9.0%gcc@9.3.0+caffe2+cuda+cudnn+distributed+fbgemm+gloo+kineto+magma~metal+mkldnn+mpi+nccl+nnpack+numa+numpy+onnx_ml+openmp+qnnpack~rocm+tensorpipe+valgrind+xnnpack cuda_arch=35 patches=e37afffe45cf7594c22050109942370e49983ad772d12ebccf508377dc9dcfc9 arch=linux-ubuntu20.04-skylake
```
As you can see, Spack has chosen to assign a random value to `cuda_arch` instead of disabling `+cuda`. With the following `packages.yaml`, Spack still tries to install `py-torch+cuda`:
```yaml
packages:
  py-torch:
    variants: ~cuda
```
```console
$ spack solve py-torch
==> Best of 8 considered solutions.
==> Optimization Criteria:
  Priority  Criterion                                 Value
  1         deprecated versions used                      0
  2         version weight                                0
  3         number of non-default variants (roots)        2
  4         multi-valued variants                        -1
  5         preferred providers for roots                 0
  6         number of non-default variants (non-roots)    1
  7         preferred providers (non-roots)             130
  8         compiler mismatches                           0
  9         version badness                              60
  10        count of non-root multi-valued variants      -5
  11        non-preferred compilers                       0
  12        target mismatches                             0
  13        non-preferred targets                         0

py-torch@1.9.0%gcc@9.3.0+caffe2+cuda+cudnn+distributed+fbgemm+gloo+kineto+magma~metal+mkldnn+mpi+nccl+nnpack+numa+numpy+onnx_ml+openmp+qnnpack~rocm+tensorpipe+valgrind+xnnpack cuda_arch=53 patches=e37afffe45cf7594c22050109942370e49983ad772d12ebccf508377dc9dcfc9 arch=linux-ubuntu20.04-skylake
```
I believe this is because dependencies like `magma`/`cudnn`/`nccl` also depend on CUDA. If I explicitly disable cuda from the command line, it works:
```console
$ spack solve py-torch~cuda
==> Best of 9 considered solutions.
==> Optimization Criteria:
  Priority  Criterion                                 Value
  1         deprecated versions used                      0
  2         version weight                                0
  3         number of non-default variants (roots)        2
  4         multi-valued variants                        -1
  5         preferred providers for roots                 0
  6         number of non-default variants (non-roots)    1
  7         preferred providers (non-roots)             130
  8         compiler mismatches                           0
  9         version badness                              60
  10        count of non-root multi-valued variants      -5
  11        non-preferred compilers                       0
  12        target mismatches                             0
  13        non-preferred targets                         0

py-torch@1.9.0%gcc@9.3.0+caffe2~cuda~cudnn+distributed+fbgemm+gloo+kineto+magma~metal+mkldnn+mpi+nccl+nnpack+numa+numpy+onnx_ml+openmp+qnnpack+rocm+tensorpipe+valgrind+xnnpack cuda_arch=none patches=e37afffe45cf7594c22050109942370e49983ad772d12ebccf508377dc9dcfc9 arch=linux-ubuntu20.04-skylake
```
Also, if I use:
```yaml
packages:
  all:
    variants: ~cuda
```
It works:
```console
$ spack solve py-torch
==> Best of 9 considered solutions.
==> Optimization Criteria:
  Priority  Criterion                                 Value
  1         deprecated versions used                      0
  2         version weight                                0
  3         number of non-default variants (roots)        2
  4         multi-valued variants                        -1
  5         preferred providers for roots                 0
  6         number of non-default variants (non-roots)    2
  7         preferred providers (non-roots)             130
  8         compiler mismatches                           0
  9         version badness                              60
  10        count of non-root multi-valued variants      -4
  11        non-preferred compilers                       0
  12        target mismatches                             0
  13        non-preferred targets                         0

py-torch@1.9.0%gcc@9.3.0+caffe2~cuda~cudnn+distributed+fbgemm+gloo+kineto+magma~metal+mkldnn+mpi~nccl+nnpack+numa+numpy+onnx_ml+openmp+qnnpack~rocm+tensorpipe+valgrind+xnnpack cuda_arch=none patches=e37afffe45cf7594c22050109942370e49983ad772d12ebccf508377dc9dcfc9 arch=linux-ubuntu20.04-skylake
```

### Information on your system

* **Spack:** 0.8.17-20938-ea2d4b05bc
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-skylake
* **Concretizer:** original

### Additional information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [x] I have run the failing commands in debug mode and reported the output

**Comment by adamjstewart** [comment]
Not sure if #24736 affects this at all.

**Comment by haampie** [comment]
@alalazo here's negative values for multi-valued variants too is what I noticed

**Comment by alalazo** [comment]
Negative values there are just maximizations instead of minimizations. We can format output better but I think `clingo` internally always minimizes. On a maximization it minimizes on the opposite number, which is what is shown here.

**Comment by alalazo** [comment]
> As you can see, Spack has chosen to assign a random value to cuda_arch instead of disabling +cuda

This first case without `packages.yaml` seems because we have a conflict in the package's default values that clingo has to work around to find an optimal solution. The default on `linux` is:
```python
py-torch+cuda cuda_arch=none
```
which is a conflict because of this `py-torch` directive:
```python
conflicts('cuda_arch=none', when='+cuda',
              msg='Must specify CUDA compute capabilities of your GPU, see '
              'https://developer.nvidia.com/cuda-gpus')
```
For the ASP based solver it's ~equivalent~ better (because of the reasons in the next comment), given the current rules, to assign an arbitrary `cuda_arch` or to turn cuda off. Here I'd try to have a consistent state for the package.

**Comment by alalazo** [comment]
> I believe this is because dependencies like magma/cudnn/nccl also depend on CUDA.

That's correct. The current semantics for preferences in `packages.yaml` is to make `~cuda` the new default for the variant. The issue with that is having non-consistent variant defaults. The ASP-based solver has to choose between:
- Keeping `~cuda` and flipping 3 variant values from their default values, or
- Using `+cuda` and changing just `cuda_arch`

Since having 2 deviations from the defaults is better than having 3, clingo is happy to discard the soft preference in `packages.yaml`. Honestly I am not sure if we should consider this an issue or rather something that needs to be fixed in the package recipe. Part of the problem is that there are multiple variants which are not orthogonal and with default values that are conflicting.

If we want to have a hierarchy of variant defaults (minimize the number of preferences from packages.yaml that are not respected _and then_ account for defaults in package.py) that will possibly require more optimization targets, which I am not sure we want.

**Comment by alalazo** [comment]
@tgamblin @becker33 See :point_up: for an interesting use case.

**Comment by alalazo** [comment]
As a workaround one can specify:
```yaml
packages:
  py-torch:
    variants: "~cuda~magma~cudnn"
```
so that `~magma` and `~cudnn` are defaults too.

**Comment by adamjstewart** [comment]
> Since having 2 deviations from the defaults is better than having 3, clingo is happy to discard the soft preference in packages.yaml. Honestly I am not sure if we should consider this an issue or rather something that needs to be fixed in the package recipe. Part of the problem is that there are multiple variants which are not orthogonal and with default values that are conflicting.

Could we instead consider defaults in `packages.yaml` to be hard preferences instead of soft preferences? That is, if it is physically possible to build something with the settings in `packages.yaml`, we do it without questions?

**Comment by alalazo** [comment]
> That is, if it is physically possible to build something with the settings in packages.yaml, we do it without questions?

We could, but I'm not sure we should. Doing something like that means adding two additional targets to the list of our 13 targets. I think it will have some impact on the time it takes to concretize (but need to measure to be sure). Also, it won't solve these kind of issues, just add another layer to it. For instance you can add conflicting variants to package `a` and package `b` in `packages.yaml`. If they're both pulled in as transitive dependencies you are basically in the same situation.

**Comment by alalazo** [comment]
If we don't modify the current algorithm, we should probably document more explicitly this behavior and try to clarify what it means to set a variant preference in `packages.yaml`.

**Comment by alalazo** [comment]
Added to the possible discussions here https://github.com/spack/spack/wiki/Telcon:-2021-10-13

**Comment by alalazo** [comment]
Still a bug. I'm removing this one from point releases since I think there will be semantic changes involved in fixing it. The next v0.18.0 seems a better target for something like that.

**Comment by alalazo** [comment]
This was solved in v0.18.0 by a combination of:
1. Conditional variants
2. Reworking `CudaPackage` to have `cuda_arch` conditional to `+cuda`

**Comment by adamjstewart** [comment]
The specific case of CUDA has been solved, but I still notice Spack ignoring packages.yaml variant preferences in my environment unless I copy them to spack.yaml. Specifically, with opencv.
